### PR TITLE
Merge Jenkins configuration to master (production)

### PR DESCRIPTION
**Pull request by shapi78**
This PR should merge all changes we made in order to configure Jenkins master and slaves

### CODE REVIEW

File: Puppetfile
Comment (antgel) [inline Puppetfile]: Please don't commit commented code. Why is this being removed?

File: Puppetfile
Comment (antgel) [inline Puppetfile]: This looks like a mistake.

File: hieradata/role/jenkins-slave.yaml
Comment (antgel) [inline hieradata/role/jenkins-slave.yaml]: As discussed, the only Microsoft dev things we need are:`visualstudio2017enterprise visualstudio2017-workload-netweb`. Please remove `microsoft-build-tools netfx-4.6.1-devpack`. I think this is correct, we should try. Please also add `nodejs-lts`.

File: hieradata/role/jenkins-slave.yaml
Comment (antgel) [inline hieradata/role/jenkins-slave.yaml]: TODOs in code don't create action items, please remove and open a GitHub Issue for this.

File: hieradata/role/jenkins-slave.yaml
Comment (antgel) [inline hieradata/role/jenkins-slave.yaml]: Where are these users used?

File: hieradata/role/jenkins-slave.yaml
Comment (antgel) [inline hieradata/role/jenkins-slave.yaml]: These aren't "sample rules", they're genuine. Please remove comment.

File: hieradata/role/jenkins-slave.yaml
Comment (antgel) [inline hieradata/role/jenkins-slave.yaml]: This looks like a type (33891 vs 3389). So I guess it doesn't work, and it's not necessary?

File: hieradata/role/jenkins.yaml
Comment (antgel) [inline hieradata/role/jenkins.yaml]: Why is this `psick::base::linux_class` with a `role` prefix not a `psick` prefix? @alvagante Comments?

File: site/role/manifests/ssh.pp
Comment (antgel) [inline site/role/manifests/ssh.pp]: 1. Again, I don't understand how `ssh` can be a `role`.
2. Please fix tabs to spaces and set your editor accordingly. @alvagante Do you think we should introduce an `.editorconfig` file?

**Review by antgel — CHANGES_REQUESTED** [review]
Few points from me, @alvagante can you also review please?

File: hieradata/role/jenkins.yaml
Comment (alvagante) [reply hieradata/role/jenkins.yaml]: That’s correct in general terms. 
We specify the class name we want to include , so it can be psick::ssh or openssh (using an existing openssh module) or profile::ssh , or even role::ssh even if in puppet world this is not a proper term, since i assume this looks more like a profile to manage ssh

File: hieradata/role/jenkins-slave.yaml
Comment (alvagante) [reply hieradata/role/jenkins-slave.yaml]: On all the nodes, windows or linux, which include the psick::users class

File: site/role/manifests/jenkins_slave.pp
Comment (alvagante) [inline site/role/manifests/jenkins_slave.pp]: As noted in an earlier comment, in Puppet world this is a profile, should be called profile::jenkins_slave ). Tecnically speaking is the same, but given the established naming conventions, this could bring some confusion for newcomers

File: site/role/manifests/ssh.pp
Comment (alvagante) [reply site/role/manifests/ssh.pp]: +1 for .editorconfig . I think I should place it upstream too

File: hieradata/role/jenkins.yaml
Comment (antgel) [reply hieradata/role/jenkins.yaml]: We want to use the proper terms. This profile is designed to enable multiple ssh keys on the server. Would you suggest a name like `profile:ssh-multiple`? Is this implementation something that could or should go into upstream?

File: site/role/manifests/jenkins_slave.pp
Comment (alvagante) [inline site/role/manifests/jenkins_slave.pp]: This is not used here

File: site/role/manifests/ssh.pp
Comment (alvagante) [inline site/role/manifests/ssh.pp]: What’s the reason to replicate here the define from psick? If there are small things to add like known hosts management this can be done directly upstream

File: hieradata/role/jenkins-slave.yaml
Comment (antgel) [reply hieradata/role/jenkins-slave.yaml]: We don't currently use this, we use `ubuntu` and `Administrator` users on Linux and Windows respectively. What would be great, would be to somehow link this to the openvpn, so that every user could have an openvpn configuration generated as well as the existing functionality. Obviously not for now, but a future improvement?

File: site/role/manifests/jenkins_slave.pp
Comment (alvagante) [inline site/role/manifests/jenkins_slave.pp]: This should be managed on the module hiera data with different values for different os

File: hieradata/role/jenkins.yaml
Comment (alvagante) [reply hieradata/role/jenkins.yaml]: You can have your profile ( do not use - in class names) or we can do it upstream, with additions to https://github.com/example42/puppet-psick/blob/master/manifests/users.pp

File: site/role/manifests/ssh.pp
Comment (alvagante) [inline site/role/manifests/ssh.pp]: A generic define as this one is supposed to be should not have user specific defaults

File: hieradata/role/jenkins-slave.yaml
Comment (alvagante) [inline hieradata/role/jenkins-slave.yaml]: Note that the order of inclusion of the classes defined here iis pre -> base -> profiles.
So here you include the role::jenkins_slave class in base, before the ones defined before in profiles (which maybe is actually desired)

File: hieradata/role/jenkins-slave.yaml
Comment (alvagante) [reply hieradata/role/jenkins-slave.yaml]: FYI, to manage host key verification issues with Jenkins, you can use something like this:
https://github.com/example42/psick-hieradata/blob/jenkinsdata/data/nodes/jenkins.foss.psick.io.yaml#L88

File: hieradata/role/jenkins-slave.yaml
Comment (alvagante) [reply hieradata/role/jenkins-slave.yaml]: So, looking at the hieradata here, psick::users is used only on Linux nodes, so here those users are not created, and actually should not probably be in a role specific file.
All the psick params with an _hash suffix are looked up in a hiera deep merge mode, so you can specify common users for all the nodes in common.yaml and have role specific ones here, if needed

File: hieradata/role/jenkins-slave.yaml
Comment (alvagante) [reply hieradata/role/jenkins-slave.yaml]: These are currently not used as the line here is commented:
https://github.com/graduway/psick/pull/34/files#diff-544566d55af404229cf6a2c9b73cdd12R14

Important note about the psick::pre|base|profile::linux|windows_classes parameters, which manage classification via psick: the keys are looked uo in deep merge mode, so it you want to disable in a specific node or role a class which is included in common.yaml (it's not this case, anyway) you should override an existing subkey (like firewall, or an other name) with an empty string. Some more details on the whole classification logic are here: https://www.example42.com/2018/01/15/classification_and_first_run_mode_with_psick/

File: hieradata/role/jenkins-slave.yaml
Comment (alvagante) [inline hieradata/role/jenkins-slave.yaml]: note the apsick here, which basically configures an unexisting class so it does nothing

File: site/role/manifests/ssh.pp
Comment (alvagante) [inline site/role/manifests/ssh.pp]: This define should stay in a file called role/manifest/ssh/ssh_keys.pp

File: site/role/manifests/ssh.pp
Comment (alvagante) [inline site/role/manifests/ssh.pp]: @shapi78 @antgel Generally what is done in this role::ssh class can probably be already done with existing psick resources or with  more or less small additions, but it can also be sanely done in a custom profile class, renaming role->profile here and in the hieradata. Or if you are in a hurry and this works, you can keep this as is.

File: Puppetfile
Comment (shapi78) [reply Puppetfile]: We don't have sensu monitor on our system, I'll remove it

Comment (shapi78) [reply Puppetfile]: If we'll plan to use sensu monitoring, I'll add it

File: Puppetfile
Comment (shapi78) [reply Puppetfile]: The module name was changed on puppetforge, we use this module to reboot windows after puppet installations

File: hieradata/role/jenkins-slave.yaml
Comment (shapi78) [reply hieradata/role/jenkins-slave.yaml]: I think we might need it to create jenkins user on Windows Jenkins-Slave machine

Comment (antgel) [reply hieradata/role/jenkins-slave.yaml]: But there's no jenkins user in the list.

File: site/role/manifests/ssh.pp
Comment (antgel) [reply site/role/manifests/ssh.pp]: I'd like to see if we can do it with existing resources (with a tweak here or there). @alvagante Can you come up with an action plan? Yes, we'd like it soon, but a day here or there is worth it to do it properly.

File: hieradata/role/jenkins.yaml
Comment (antgel) [reply hieradata/role/jenkins.yaml]: @alvagante I'll try to be clearer. This profile is intended to enable multiple ssh keys on the Jenkins slave, which are deploy keys for two different GitHub repositories. It's not intended for managing ssh keys for real users (although perhaps the distinction is irrelevant). What do you suggest?

File: site/role/manifests/ssh.pp
Comment (antgel) [reply site/role/manifests/ssh.pp]: Can you add upstream and we'll rebase? We're a bit out of sync with upstream at the moment. I'd suggest something like:
```# Editor configuration, see http://editorconfig.org
root = true

[*]
charset = utf-8
end_of_line = lf
indent_style = space
indent_size = 2
insert_final_newline = true
trim_trailing_whitespace = true

[*.md]
max_line_length = off
trim_trailing_whitespace = false
```

File: site/role/manifests/ssh.pp
Comment (alvagante) [reply site/role/manifests/ssh.pp]: The psick::users class can iterate over the  psick::users::managed define https://github.com/example42/puppet-psick/blob/jenkins/manifests/users/managed.pp which already allows, for each user to manage via params, configurable on Hiera, ssh keys and known hosts source (but not the content, at the moment, but that's easy to add).
I have to test it on Windows though, as I think it might need some tweaks there.
Let me know if to implement the above things and if that's actually all that you need.

Comment (antgel) [reply site/role/manifests/ssh.pp]: @alvagante Green light from me. When it's ready, let us know and we'll finally rebase our psick on upstream. Sane? 
@shapi78 I think you should rebase this branch to drop the ssh key commit, then at least we can get it into master. Sane?

Comment (antgel) [reply site/role/manifests/ssh.pp]: @alvagante To clarify, this will let us put two different ssh keys in the homedir of the jenkins user on our Windows Jenkins slave? So that we can use one for the `web` repo and the other for the `frontend` repo?

Comment (alvagante) [reply site/role/manifests/ssh.pp]: @antgel you need 2 different keys in (4) different files I suppose.
Currently, for a given user, psick::users::managed just manages id_rsa and id_rsa.pub but maybe at this point it's better just to create a generic define like psick::ssh::key which just manages the keys and not the user (so somehow more similar to the current one in role).


Anyway, since the need is very limited and confined (keys for web and frontend repo for jenkins user on windows) you could manage the relevant files resource directly in role::jenkins_slave (renamed or to to profile) without the extra role::ssh stuff

Comment (antgel) [reply site/role/manifests/ssh.pp]: Fine. @shapi78 Please fix the ssh commit as per @alvagante and my recommendations. Indeed, why don't we just put this resource inside the existing role::jenkins_slave?

File: site/role/manifests/ssh.pp
Comment (alvagante) [reply site/role/manifests/ssh.pp]: Added, will be merged to psick control repo soon.
Note that current psick has already the hieradata moved to a separated module.
So when you merge you'll have to manage hiera.yaml and Puppetfile as needed (can help with that if needed)

Comment (antgel) [reply site/role/manifests/ssh.pp]: Yes, I know we spoke about it, but I'd still like to set a meeting to do it. @shapi78 Please install editorconfig plugin in your editor of choice. :)

**Comment by shapi78** [comment]
Fixed jenkins.yaml role to work with new psick::openssh module. works locally, waiting for upstream psick::openssh fix before merge

File: hieradata/role/jenkins-slave.yaml
Comment (antgel) [inline hieradata/role/jenkins-slave.yaml]: Please remove `visualstudio2017-workload-netweb` and `visualstudio2017enterprise`, and install `webdeploy`.

File: hieradata/role/jenkins-slave.yaml
Comment (antgel) [inline hieradata/role/jenkins-slave.yaml]: Are we using these?

File: manifests/site.pp
Comment (antgel) [inline manifests/site.pp:25]: Where is `$host_name` used?

**Review by antgel — CHANGES_REQUESTED** [review]

File: manifests/site.pp
Comment (shapi78) [reply manifests/site.pp:25]: To get the Name tag from aws and set it as its hostname

Comment (antgel) [reply manifests/site.pp:25]: But I don't see that variable used anywhere in our code.

Comment (antgel) [reply manifests/site.pp:25]: (And hostname is one word, for what it's worth.)

Comment (shapi78) [reply manifests/site.pp:25]: `set_hostname() {
  instance_id=$(curl 169.254.169.254/latest/meta-data/instance-id)
  host_name=${hostname}-$${instance_id}
  domain=${environment}.graduway.com
  fqdn=$${host_name}.$${domain}
  hostnamectl set-hostname "$${fqdn}"
  echo 127.0.0.1 $${host_name} $${fqdn} >> /etc/hosts
}`
it uses it during boot proccess using terraform, so you configure the name tag and it causes the linux host name to be aligned with that

Comment (antgel) [reply manifests/site.pp:25]: Cool. I'd add a small comment, explaining that the variable is used in Terraform. Can be on a future branch.

**Review by antgel — APPROVED** [review]

**Pull request merged**
